lib/TPP/IR/StructuredOpMatcher.cpp

```diff
@@ -201,6 +201,80 @@ bool tpp::structured_match::WithSingleOpImpl::withSingleOpImpl(
   return true;
 }
 
+// FIXME: This is a generalization of the method above and will eventually replace
+// the matcher for both no-op (yield) and one op (add, max).
+bool tpp::structured_match::withOpChainImpl(
+    Region *region, Operation *op, SmallVectorImpl<Value> *capturedOperands,
+    SmallVectorImpl<TypeCheckFunc>& typeChecks) {
+
+  // Numer of ops includes yield
+  ptrdiff_t numOps = typeChecks.size();
+
+  // Basic checks
+  if (!isa<linalg::GenericOp>(op))
+    return false;
+  auto linalgOp = cast<linalg::GenericOp>(op);
+  if (!region->hasOneBlock())
+    return false;
+  auto& block = region->front();
+  if (std::distance(block.begin(), block.end()) != (numOps + 1))
+    return false;
+  if (linalgOp.getNumDpsInits() != 1)
+    return false;
+
+  // Add generic arguments to the list of chained values
+  llvm::SmallSetVector<Value, 4> chainedValues;
+  for (auto arg: block.getArguments()) {
+    chainedValues.insert(arg);
+  }
+
+  // Check on the inner chain of operations in the right order.
+  // Make sure all operands are used and chained
+  for (auto [check, innerOp]: llvm::zip_first(typeChecks, block.getOperations())) {
+    // Must be right op in right order
+    if (!check(&innerOp))
+      return false;
+
+    // At least one operand must come from args or a previous op
+    bool consumesValueFromChain = false;
+    for (auto operand: innerOp.getOperands()) {
+      if (chainedValues.contains(operand)) {
+        // First add to the captured
+        if (capturedOperands) {
+          if (auto ba = dyn_cast<BlockArgument>(operand)) {
+            if (ba.getParentBlock() == linalgOp.getBlock()) {
```

Comment:
nit: it'd be easier to read with reduced nesting depth

Reply:
Yeah, this was a result of iteration development, simplifying it.